File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var solve = function(nums,ans,i){
    if(i==nums.length){
        ans.push([...nums]);
        return;
    }
    for(let j=i;j<nums.length;j++){
        [nums[i],nums[j]]=[nums[j],nums[i]]
        solve(nums,ans,i+1);
        [nums[i],nums[j]]=[nums[j],nums[i]];
    }
    
}
var permute = function(nums) {
    let ans=[];
    solve(nums,ans,0);
    return ans;
};

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>>ans;
        vector<int>temp;
        vector<bool>visited(nums.size(),0);
        solve(nums,ans,temp,visited);
        return ans;
    }
    void solve(vector<int>& nums,vector<vector<int>>&ans,vector<int>temp,vector<bool>visited){
        if(temp.size()==nums.size()){
            ans.push_back(temp);
        }
        for(int i=0;i<nums.size();i++){
            if(visited[i]==0){
                temp.push_back(nums[i]);
                visited[i]=1;
                solve(nums,ans,temp,visited);
                visited[i]=0;
                temp.pop_back();
            }
        }
    }
};